### PR TITLE
<filesystem>: Fix directory_iterator returning "dot" and "dotdot" entries

### LINKED ISSUE #494: <filesystem>: directory_iterator including "." and ".." on network drives
**Describe the bug**
Our directory enumeration machinery assumes `.` and `..` will be the first directory entries returned from FindFirstFile, but this doesn't appear to be the case in some network share scenarios.

This is a dual of DevCom-903816 which has additional details.

**Command-line test case**
STL version (git commit or Visual Studio version): Visual Studio 2019 version 16.5

Create a directory and create a file #1.txt on Windows 10 host. Share this directory to Windows 10 client.

```
#include <iostream>
#include <filesystem>
int main()
{
    try {
        namespace fs = std::filesystem;
        fs::path dir = "//ServerHost/D$/work/dirTest";
        dir = absolute(dir);

        for (auto const& di : fs::directory_iterator(dir)) {
            if (di.is_directory()) {
                std::wcout << "is_directory:  ";
            }
            else {
                std::wcout << "!is_directory: ";
            }
            std::wcout << di.path().native() << std::endl;
        }
    }
    catch(std::filesystem::filesystem_error &e) {
        std::wcout << e.what();
    }
    return 0;
}
```

**Actual behavior**
```
!is_directory: \\ServerHost\D$\work\dirTest1\#1.txt
is_directory:  \\ServerHost\D$\work\dirTest\.
is_directory:  \\ServerHost\D$\work\dirTest\..
```

**Expected behavior**
```
!is_directory: \\ServerHost\D$\work\dirTest\#1.txt
```

### PR DESCRIPTION
**Pull request by JeanPhilippeKernel**
# Description

Fixes #494.

# Checklist

Be sure you've read README.md and understand the scope of this repo.

If you're unsure about a box, leave it unchecked. A maintainer will help you.

- [x] Identifiers in product code changes are properly `_Ugly` as per
  https://eel.is/c++draft/lex.name#3.1 or there are no product code changes.
- [x] The STL builds successfully and all tests have passed (must be manually
  verified by an STL maintainer before automated testing is enabled on GitHub,
  leave this unchecked for initial submission).
- [x] These changes introduce no known ABI breaks (adding members, renaming
  members, adding virtual functions, changing whether a type is an aggregate
  or trivially copyable, etc.).
- [x] These changes were written from scratch using only this repository,
  the C++ Working Draft (including any cited standards), other WG21 papers
  (excluding reference implementations outside of proposed standard wording),
  and LWG issues as reference material. If they were derived from a project
  that's already listed in NOTICE.txt, that's fine, but please mention it.
  If they were derived from any other project (including Boost and libc++,
  which are not yet listed in NOTICE.txt), you *must* mention it here,
  so we can determine whether the license is compatible and what else needs
  to be done.

### CODE REVIEW

**Comment by StephanTLavavej** [comment]
Thanks! I've edited your PR description to say "Fixes #494." We recommend following this pattern in the future, as it'll link your PR to the issue that it fixes. See [GitHub's documentation](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) "Linking a pull request to an issue using a keyword".

File: stl/inc/filesystem [hunk 1]
```diff
@@ -2489,12 +2489,18 @@ namespace filesystem {
         _NODISCARD static __std_win_error _Advance_and_reset_if_no_more_files(shared_ptr<_Dir_enum_impl>& _Ptr) {
             auto& _Impl = *_Ptr;
             __std_fs_find_data _Data;
-            const auto _Error = __std_fs_directory_iterator_advance(_Impl._Dir._Handle, &_Data);
-            if (_Error == __std_win_error::_Success) {
-                _Impl._Refresh(_Data); // can throw
-            } else if (_Error == __std_win_error::_No_more_files) {
+            auto _Error = __std_fs_directory_iterator_advance(_Impl._Dir._Handle, &_Data);
```
Comment (BillyONeal) [inline stl/inc/filesystem]: Proposal:

```
do {
auto _Error = __std_fs_directory_iterator_advance(_Impl._Dir._Handle, &_Data);
if (_Error != __std_win_error::_Success) {
_Ptr.reset();
return _Error;
}
} while (_Is_dot_or_dotdot(_Data));
return __std_win_error::_Success;
```
?

**Comment by JeanPhilippeKernel** [comment]
> Thanks! I've edited your PR description to say "Fixes #494." We recommend following this pattern in the future, as it'll link your PR to the issue that it fixes. See [GitHub's documentation](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) "Linking a pull request to an issue using a keyword".

Thanks a lot !! after sending the PR at 2:30 am, being in bed I thought I had forgotten something, it was actually this 😂😂😂

File: stl/inc/filesystem (hunk 1 above)
Comment (JeanPhilippeKernel) [reply stl/inc/filesystem]: hmm yeah, it seems more  perfect and clear than mine, let me test it

Comment (JeanPhilippeKernel) [reply stl/inc/filesystem]: Great ! it works very fine !! Thanks :)

**Comment by JeanPhilippeKernel** [comment]
Yes, I confirm that it works fine for both local and network drives ! 
Dot and dod-dot are correctly skipped now

File: stl/inc/filesystem [hunk 2]
```diff
@@ -2489,15 +2489,19 @@ namespace filesystem {
...
             __std_fs_find_data _Data;
-            const auto _Error = __std_fs_directory_iterator_advance(_Impl._Dir._Handle, &_Data);
-            if (_Error == __std_win_error::_Success) {
-                _Impl._Refresh(_Data); // can throw
-            } else if (_Error == __std_win_error::_No_more_files) {
-                _Ptr.reset();
-                return __std_win_error::_Success;
-            }
+            do {
+                auto _Error = __std_fs_directory_iterator_advance(_Impl._Dir._Handle, &_Data);
```
Comment (CaseyCarter) [inline stl/inc/filesystem]:
```suggestion
                const auto _Error = __std_fs_directory_iterator_advance(_Impl._Dir._Handle, &_Data);
```

**Comment by JeanPhilippeKernel** [comment]
Great , thanks, @BillyONeal  I was trying to squash commit of this branch, to make it more readable

**Comment by StephanTLavavej** [comment]
The logic also makes sense to me. Thank you for this runtime correctness fix!

**Comment by JeanPhilippeKernel** [comment]
Sorry for having this "long commit history" on this PR:(

**Comment by BillyONeal** [comment]
Nothin to be sorry about :).

**Comment by BillyONeal** [comment]
Thanks for your contribution! This should ship in Visual Studio 2019 version 16.7 preview 1.

**Pull request merged**